### PR TITLE
v3.1.1

### PR DESCRIPTION
**Pull request by betzrhodes**
- Added backoff when 429 errors encountered
- Make user callbacks executed synchronously
- Move statics to constants
- Some (internal) refactoring

### CODE REVIEW

File: Firebase.agent.lib.nut [hunk 1]
```diff
@@ -24,10 +24,11 @@
 
 class Firebase {
     // Library version
-    static VERSION = "3.0.0";
+    static VERSION = "3.0.1";
```
Comment (ppetrosh) [inline Firebase.agent.lib.nut]: I would probably make it 3.1.0 as it's a new feature added.

File: Firebase.agent.lib.nut [hunk 2]
```diff
@@ -641,24 +655,57 @@ class Firebase {
...
+                    local now = time();
+                    // Too many requests, set _tooManyReqTimer to prevent more requests to FB
+                    if (_tooManyReqTimer == false) {
+                        // This is the first 429 we have seen set a default timeout
+                        _tooManyReqTimer = now + DEFAULT_BACK_OFF_TIMEOUT_SEC;
+                    } else if (_tooManyReqTimer <= now) {
+                        // Firebase is still overwhelmed after first timeout expired, 
+                        // Let's block requests for longer to let FB recover
+                        _tooManyReqCounter += 1
+                        _tooManyReqTimer = now + (DEFAULT_BACK_OFF_TIMEOUT_SEC * _tooManyReqCounter);
```
Comment (ppetrosh) [inline Firebase.agent.lib.nut]: `_tooManyReqTimer = now + (DEFAULT_BACK_OFF_TIMEOUT_SEC * _tooManyReqCounter++);` and the prev line can be removed

File: Firebase.agent.lib.nut [hunk 3]
```diff
@@ -641,24 +655,57 @@ class Firebase {
...
+        // Only send request if we haven't received a 429 error recently
+        if (_tooManyReqTimer == false || _tooManyReqTimer <= time()) {
+            if (usePromise) {
+                return _createRequestPromise(request);
+            } else {
+                return _sendRequest(request, callback);
+            }
         } else {
-            return _sendRequest(request, callback);
+            local error = "ERROR: Too many requests to Firebase, try request again after " + _tooManyReqTimer;
```
Comment (ppetrosh) [inline Firebase.agent.lib.nut]: Will a timestamp be printed here? Will it be more user friendly to say "in ... sec"?

File: Firebase.agent.lib.nut [hunk 4]
```diff
@@ -641,24 +655,57 @@ class Firebase {
...
-    // use Promise if promise libary included and callback != null
     function _processResponse(request, callback) {
-        if (_promiseIncluded && callback == null) {
-            return _createRequestPromise(request);
+        // Use Promise if promise libary included and callback != null
+        local usePromise = (_promiseIncluded && callback == null);
+
+        // Only send request if we haven't received a 429 error recently
+        if (_tooManyReqTimer == false || _tooManyReqTimer <= time()) {
+            if (usePromise) {
```
Comment (ppetrosh) [inline Firebase.agent.lib.nut]: `return usePromise ? _createRequestPromise(request) : _sendRequest(request, callback);` is shorter

File: Firebase.agent.lib.nut [hunk 5]
```diff
@@ -641,24 +655,53 @@ class Firebase {
...
             }
         }
     }
 
-    // use Promise if promise libary included and callback != null
     function _processResponse(request, callback) {
-        if (_promiseIncluded && callback == null) {
-            return _createRequestPromise(request);
+        // Use Promise if promise libary included and callback != null
+        local usePromise = (_promiseIncluded && callback == null);
```
Comment (ppetrosh) [inline Firebase.agent.lib.nut:681]: I know it's not your code, but I'm wondering where is this check for callback comes from? Why can't the callback be executed from a promise? Do you have an idea?

File: Firebase.agent.lib.nut [hunk 6]
```diff
@@ -641,24 +655,53 @@ class Firebase {
...
+                        _tooManyReqTimer = now + (DEFAULT_BACK_OFF_TIMEOUT_SEC * _tooManyReqCounter++);
+                    }
+                    // Pass error to callback
+                    onError("Error " + res.statuscode);
                 } else if (typeof data == "table" && "error" in data) {
+                    _tooManyReqTimer = false;
                     local error = data ? data.error : null;
                     onError(error);
                 } else {
+                    _tooManyReqTimer = false;
```
Comment (ppetrosh) [inline Firebase.agent.lib.nut:669]: I believe we also need to handle 503 (and may be others) the same way we do for 429. But we can probably add this later if needed.

File: Firebase.agent.lib.nut (hunk 5 above)
Comment (betzrhodes) [reply Firebase.agent.lib.nut:681]: The library should function if the Promise library is not included. As I understand it if you include a callback that will be used, if you don't include a callback but have used the Promise library the CRUD methods will return a Promise.  I have not fully tested this.

File: README.md [hunk 7]
```diff
@@ -1,8 +1,8 @@
-# Firebase v3.0.0
+# Firebase v3.1.0
```
Comment (ppetrosh) [inline README.md]: Shouldn't we update the documentation accordingly to describe the new behavior?

File: Firebase.agent.lib.nut (hunk 5 above)
Comment (ppetrosh) [reply Firebase.agent.lib.nut:681]: I see. Thanks.

File: Firebase.agent.lib.nut (hunk 6 above)
Comment (betzrhodes) [reply Firebase.agent.lib.nut:669]: I can add 503 to the check pretty easily.

**Review by ppetrosh — APPROVED** [review]

**Pull request merged**